File: core/src/main/java/com/serious/service/ServiceAddress.java
package com.serious.service;
/*
 * @COPYRIGHT (C) 2016 Andreas Ernst
 *
 * All rights reserved
 */

import lombok.EqualsAndHashCode;
import lombok.Getter;
import org.springframework.cloud.client.ServiceInstance;

import java.net.URI;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * @author Andreas Ernst
 */
@Getter
public class ServiceAddress {
    // static methods

    private static List<ServiceAddress> extractAddresses(String channels) {
        return  Arrays.stream(channels.split(","))
                .map(channel -> {
                    int lparen = channel.indexOf("(");
                    int rparen = channel.indexOf(")");

                    String protocol = channel.substring(0, lparen);
                    URI uri = URI.create(channel.substring(lparen + 1, rparen));

                    return new ServiceAddress(protocol, uri);
                })
                .collect(Collectors.toList());
    }

    private static ServiceAddress getAddress(String channel, ServiceInstance instance) {
        List<ServiceAddress> addresses = extractAddresses(instance.getMetadata().get("channels"));

        return addresses.stream().filter(address -> address.getChannel().equals(channel)).findFirst().orElse(null);
    }

    // instance data

    public String channel;
    private URI uri;
    public ServiceInstance serviceInstance;

    // constructor

    public ServiceAddress() {
    }

    public ServiceAddress(String channel, URI uri) {
        this.channel = channel;
        this.uri = uri;
    }

    public ServiceAddress(String channel, ServiceInstance instance) {
        this.channel = channel;
        this.serviceInstance = instance;
        this.uri = getAddress(channel, instance).getUri();//instance.getUri();
    }

    // override Object


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ServiceAddress that)) return false;
        return Objects.equals(channel, that.channel) && Objects.equals(uri, that.uri);
    }

    @Override
    public int hashCode() {
        return Objects.hash(channel, uri);
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();

        if ( channel != null) {
            builder
                    .append(this.channel).append("(")
                    .append(this.uri.toString()).append(")");
        }
        else builder.append("-");

        return builder.toString();
    }
}

File: core/src/main/java/com/serious/service/ServiceAddress.kt
package com.serious.service
/*
* @COPYRIGHT (C) 2023 Andreas Ernst
*
* All rights reserved
*/

import org.springframework.cloud.client.ServiceInstance
import java.net.URI
import java.util.*
import java.util.stream.Collectors

/**
 * A <code>ServiceAddress</code> is a resolved [ChannelAddress]
 */
data class ServiceAddress(var component: String, var channel: String, var serviceInstances: List<ServiceInstance>) {
    // instance data

    val uri : List<URI>

    init {
        uri = computeURIs()
    }

    fun changed(address : ServiceAddress) : Boolean {
        if ( channel != address.channel)
            return true

        if ( uri.size != address.uri.size )
            return true

        if ( !HashSet<URI>(uri).equals(HashSet<URI>(address.uri)))
            return false

        // actually we don't care...
        //if ( serviceInstances.size != address.serviceInstances.size )
        //    return true

        // oke

        return false
    }

    // private

    private fun computeURIs() :List<URI> {
        val result = HashSet<URI>()

        for (serviceInstance in serviceInstances ) {
            val channels = serviceInstance.metadata.get("channels")!!

            for ( channel in channels.split(",".toRegex())) {
                val lparen = channel.indexOf("(")
                val rparen = channel.indexOf(")")
                val protocol = channel.substring(0, lparen)

                if (protocol == this.channel)
                    result.add(URI.create(channel.substring(lparen + 1, rparen)))
            }
        }

        return result.toList()
    }

    // override

    override fun toString(): String {
        val builder = StringBuilder()

        builder
            //.append("<").append(component).append(">")
            .append(channel).append("(")

        for ( instance in serviceInstances  )
            builder.append(instance.toString())

        builder
            .append(")")

        return builder.toString()
    }

    companion object {
        fun localAddress(componentDescriptor: ComponentDescriptor<*>) : ServiceAddress {
            return ServiceAddress(componentDescriptor.name, "local", emptyList())
        }

        fun missingAddress(component: String) : ServiceAddress {
            return ServiceAddress(component, "missing", emptyList())
        }
    }
}
